Plugins/BuildServerIntegration/AzureDevOpsIntegration/AzureDevOpsAdapter.cs

```diff
@@ -122,15 +137,41 @@ private async Task ObserveBuildsAsync(DateTime? sinceDate, bool running, IObserv
             {
                 if (_buildDefinitions == null)
                 {
-                    _buildDefinitions = await _buildDefinitionsTask.JoinAsync();
+                    try
+                    {
+                        _buildDefinitions = await _buildDefinitionsTask.JoinAsync();
+
+                        if (_buildDefinitions == null)
+                        {
+                            observer.OnCompleted();
+                            return;
+                        }
 
-                    if (_buildDefinitions == null)
+                        CacheAzureDevOps = new CacheAzureDevOps { Id = CacheKey, BuildDefinitions = _buildDefinitions };
+                    }
+                    catch (UnauthorizedAccessException)
                     {
-                        observer.OnCompleted();
+                        ManageError(_badTokenErrorMessage.Text);
                         return;
                     }
+                    catch (Exception ex)
+                    {
+                        ManageError(_genericErrorMessage.Text + ex.Message);
+                        return;
+                    }
+                }
+
+                void ManageError(string errorMessage)
+                {
+                    _apiClient = null;
+                    observer.OnNext(new BuildInfo { CommitHashList = new[] { ObjectId.WorkTreeId }, Description = errorMessage, Status = BuildInfo.BuildStatus.Failure });
+                    observer.OnCompleted();
 
-                    CacheAzureDevOps = new CacheAzureDevOps { Id = CacheKey, BuildDefinitions = _buildDefinitions };
+                    if (ProjectOnErrorKey == null || ProjectOnErrorKey != CacheKey)
+                    {
+                        ProjectOnErrorKey = CacheKey;
+                        MessageBox.Show(errorMessage, _buildIntegrationErrorCaption.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
```

Comment:
I think we could do a little better with the use of Task Dialog - we show the message (in the same way we do it now) and have two buttons: `[Fix PAT]` and `[Continue]` (the buttons text is up for a discussion).
The first button will take a user to settings where a new PAT can be inserted, and the other button basically ignores the error.

Reply:
>The first button will take a user to settings where a new PAT can be inserted

I had a look but unfortunately it seems that we have no way to open the user settings from a build integration plugin...

Reply:
Other plugins can do it:
https://github.com/gitextensions/gitextensions/blob/588fc0f5b0666855a225d9507ff17ee78f2fe27e/Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs#L225
https://github.com/gitextensions/gitextensions/blob/588fc0f5b0666855a225d9507ff17ee78f2fe27e/Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs#L60

Do you think you could spend a little more time here to see what can be done?

Reply:
Working WIP pushed...

![image](https://user-images.githubusercontent.com/460196/95745451-68a16480-0c95-11eb-9369-ca670afa3009.png)